src/lib/blkdata_svc/blk_read_tracker.cpp

```diff
@@ -18,69 +17,97 @@
 namespace homestore {
 static BlkId extract_key(const BlkTrackRecord& rec) { return rec.m_key; }
 
-static bool blkid_overlaps(const BlkId& a, const BlkId& b) {
-    if (a.get_chunk_num() != b.get_blk_num()) { return false; }
-    if (a.get_blk_num() == b.get_blk_num()) { return true; }
-    if (a.get_blk_num < b.get_blk_num()) { return ((a.get_blk_num() + a.get_nblks()) >= b.get_blk_num()); }
-    return ((b.get_blk_num() + b.get_nblks()) >= a.get_blk_num());
-}
-
-static bool merge_records(BlkTrackRecord& rec, const BlkTrackRecord& other) {
-    assert(rec.m_key == other.m_key);
-    assert(other.m_sub_record.size() == 1); // We don't support multiple elements to merge with each other
-
-    const auto& new_rec = other.m_sub_record[0];
-    rec.m_total_refcount += other.m_total_refcount;
+BlkReadTracker::BlkReadTracker() : m_pending_reads_map(s_expected_num_records, extract_key, nullptr /* access_cb */) {}
 
-    bool inserted{false};
-    auto it = rec.m_sub_record.begin();
-    while (it != rec.m_sub_record.end()) {
-        if (blkid_overlaps(it->sub_range, new_rec.sub_range)) {
-            it->ref_count += new_rec.ref_count;
-            if (it->ref_count == 0) { // No one references, removed from list
-                it = rec.m_sub_record.erase(it);
-            } else {
-                it->waiter = new_rec.waiter;
-                inserted = true;
-            }
-        }
-        ++it;
+void BlkReadTracker::merge(const BlkId& blkid, int64_t new_ref_count,
+                           const std::shared_ptr< blk_track_waiter >& waiters) {
+    if (new_ref_count == 0) {
+        assert(waiters);
+    } else {
+        assert(!waiters);
```

Comment:
Cosmetic suggestion, we can use HS_ASSERT. Also we can brief this as

HS_ASSERT(new_ref_count ? !waiters : waiter, "Invalid waiters");